### ISSUE TITLE
samtools view to fetch specific chromosome got one less read

### ISSUE THREAD

**Issue by phuang7008** [samtools/samtools#1574]
Hi HTSLIB developers,

I am using `htslib --version 1.10.2` and '1.14'
I am using Linux 2.6.32-573.26.1.el6.x86_64
I am using gcc version 8.3.1 20190311 (Red Hat 8.3.1-3) (GCC) 

I have 1000+ cram files. I developed a software that uses htslib. When I run my software with these cram files, about one thirds of my cram files have total reads produced with one less read when compare to samtools flagstat results. The same thing happens to the "samtools view -hb my.cram chr_id > chr_id.bam" command with one less read as well.

the missing read could be in any chromosome. Here I will give an example where the missing read is on chr: chrUn_JTFH01000737v1_decoy

When I used samtools view to open the cram file and using grep "chrUn_JTFH01000737v1_decoy" command, I was able to get 340 reads. However, when I used "samtools view -hb my.cram chrUn_JTFH01000737v1_decoy > chrUn_JTFH01000737v1_decoy.bam" command, I only get 339 reads.

All reads are paired reads. Here are the problematic paired reads:
A00733:398:H2V2HDSX3:1:1236:19488:17143 163     chrUn_JTFH01000737v1_decoy      1       60      17S134M =       192     361     AATCTCAGCTACTCGGGAGGCTGAGGCAGGAGAATGGCGTGAACCCGGGAGGCGGAGCTTGCAGTGAGCCGAGATGGCGCCACTGCACTCCAGCCTGGGGGGACAGACCGAGACTCTGTCTCAAAAAAGAAAAAAAAAAATACAAAAATTA    FFFFFFFFFFFF,FFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFF,,:FF    RG:Z:H2V2HDSX3.1        AS:i:134   XS:i:95 NM:i:0  XQ:i:180        MD:Z:134

A00733:398:H2V2HDSX3:1:1236:19488:17143 83      chrUn_JTFH01000737v1_decoy      192     60      91M19D60M       =       1       -361    AATCCCTTGAACCCTGGAGGTAGAGGTTGCAGTGAGCAGAGGTCATGCCACTGTACTCCAGCCTGGGCGTCACAGCAAGACTTTGTCTCAGAAAAAAAAAAAAAAGAAGAGGGACTATAGAAAGAGGCAAATTTCAGTGTAAAGAGGAATT    F:FF::FFFFFFF,,FF:F,FF,:FFFFFFFFF,FF:::,FFFFFFFFFFF:F:F:FFF:FF,FFFFFFFFFF:F,F:,F,FFFFF:FF:,FFFFFFFFFFFFFFF:,F,F::,FFF,FFFFFFFF,FFFFFFFFFFFFFFFFFFFFFFFF    RG:Z:H2V2HDSX3.1  AS:i:121 XS:i:42 NM:i:20 XQ:i:250        MD:Z:14A76^AAAAAAAAAAAAAAAAAAA60

where the missing read is:
A00733:398:H2V2HDSX3:1:1236:19488:17143 163     chrUn_JTFH01000737v1_decoy      1       60      17S134M =       192     361     AATCTCAGCTACTCGGGAGGCTGAGGCAGGAGAATGGCGTGAACCCGGGAGGCGGAGCTTGCAGTGAGCCGAGATGGCGCCACTGCACTCCAGCCTGGGGGGACAGACCGAGACTCTGTCTCAAAAAAGAAAAAAAAAAATACAAAAATTA    FFFFFFFFFFFF,FFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFF,,:FF    RG:Z:H2V2HDSX3.1        AS:i:134   XS:i:95 NM:i:0  XQ:i:180        MD:Z:134

my question is why the "samtools view -hb my.cram chrUn_JTFH01000737v1_decoy > chrUn_JTFH01000737v1_decoy.bam" got one less read? Interestingly, all the comparison I have done so far, 1/3 of them have one less read for the cram files tested (not 2, not 3 less reads).

Sorry, I am not able to share the problematic cram files. If I get a control cram file with the same issue, I will share with you right away. Thanks!

Best,

Peter Huang

**Comment by jkbonfield** [comment]
That's a perculiar one.  Is the missing record the first or last in the chromosome?  I'd find it odd for it to miss one right out of the middle.  Note using grep to count may be problematic if there are cross-chromosome read pairs.  Checking field 3 only would be more robust, for example:

```
samtools view file.cram | awk '$3 == "chrUn_JTFH01000737v1_decoy" {n++} END {print n}'
```

Is it possible you could try the latest version of the software?  `samtools --version` should report both the version number from itself and the htslib it is linked against (which is likely where the problem lies).

Also as this is likely an indexing issue, reporting the relevant lines surrounding  chrUn_JTFH01000737v1_decoy  in the index would help.  You can first work out which reference number this is (counting from 0 rather than 1) by seeing what Nth `@SQ` header line it corresponds to.  You can then zcat the .cram.crai file to find the entries starting with that reference number, plus one either side.

**Comment by jkbonfield** [comment]
Hmm, I can see *a* bug, related to the `required_fields` selection.  With 0x2f (every value up to and including CIGAR) on one query I got 1 more RNAME matching than with `required_fields=0x1f`` (ie omitting CIGAR).  I can't see why CIGAR would be linked to RNAME decoding though - it definitely shouldn't be.

Are you doing any selective decoding of SAM columns?

**Comment by jkbonfield** [comment]
I note your read that is missing starts at position 1.

I can see an off-by-one bug in how the read start and end coordinates are initialised.  When decoding the cigar string this is fine as it sets POS to 1 and END to 1+cigar_length-1 inclusive.  When cigar isn't decoded it sets END to the ref-pos field, but that's the internal 0-based ref pos and not the 1-base POS that we use for region filtering later.    An END of 0 is less than the region we're filtering against (effectively chr:1-MAX).

This minor patch fixes it for me:

```
diff --git a/cram/cram_decode.c b/cram/cram_decode.c
index e65e11d..5ff5db8 100644
--- a/cram/cram_decode.c
+++ b/cram/cram_decode.c
@@ -1777,7 +1777,7 @@ static int cram_decode_seq(cram_fd *fd, cram_container *c, cram_slice *s,
     }
 
     cr->ncigar = ncigar - cr->cigar;
-    cr->aend = ref_pos;
+    cr->aend = ref_pos+1;
 
     //printf("2: %.*s %d .. %d\n", cr->name_len, DSTRING_STR(name_ds) + cr->name, cr->apos, ref_pos); 
```

Unfortunately I can't tell whether this is a totally different bug I've found or the same one you're seeing.  Is there any way you can verify if this is it?  Thanks.

**Comment by phuang7008** [comment]
Thanks James!

I have checked 3 cram files with the same issue, so far all of them are the first read in the corresponding chromosomes that went missing.
You are right, I did use 'awk' to get the count info from samtools view command to confirm the results.

I used samtools 1.14 for testing and the result is the same with one less read for the cram file used for testing.
$ ./samtools-1.14/samtools --version
samtools 1.14
Using htslib 1.14
Copyright (C) 2021 Genome Research Ltd.

I then tried to find out what is going on with the .cram.crai files. It turns out that some of the index files have lines like the following
2188    1968    2147481681      21628350488     390     351890
2189    1       -2147483648     21628350488     390     351890
2190    1       2058    21628350488     390     351890
The negative value on column 3 is definitely not right.

I then used 'samtools index' command to regenerate the .cram.crai files and the problem is gone. I will have to find out why the original .cram.crai files have negative values on the 3rd column.

Thanks for the kind help and pointed me to the right direction.

Best,

Peter

**Comment by jkbonfield** [comment]
Do you know what produced these indices?  Does the file header have an indication of the tools ran and their version numbers maybe?

I see https://github.com/samtools/htslib/pull/85 refers to a CRAM indexing bug.  This was code added by me to fix an issue with indexing of CRAM containers that mixed alignments against multiple references into the same slice/container.  This is the case for your data (note both ref no. 2188 and 2198 are at the same file offset of 21628350488).  That commit however was in htslib 0.2.0-rc7.  So well before the official 1.0 release.

It's possible this is just ancient data.  I hope that's the case anyway!  Thanks for doing more investigation into the issue.

**Comment by phuang7008** [comment]
Hi James,

I talked to my center and was told that we were using the Dragen pipeline from Illumina v3.4.2. Dragen team is aware this issue and has this fixed in v3.7.8. Our center is currently conducting validation tests using this new version. Hopefully this off by 1 issue won't happen again. 

Thank you very much!

Best,

Peter
